### PR TITLE
Proposal for enabling AppOnly access to Microsoft Graph

### PR DESCRIPTION
**Pull request by darrelmiller**
This PR proposes a number of updates to the Graph integration with Identity.Web.
It addresses https://github.com/AzureAD/microsoft-identity-web/issues/654, plus more.

It creates two new overloads of AddMicrosoftGraph that provide full control to the user for configuring the GraphServiceClient while still abstracting away the auth configuration.

This configures a GraphServiceClient to use delegated by default
```
.AddMicrosoftGraph(authProvider => new GraphServiceClient(authProvider), initialScopes);
```
In order to make the occasional appOnly request the `WithAppOnly()` method can be used,

```
graphServiceClient.Users[userId].Request().WithAppOnly().GetAsync();
```

This configures a HTTP client to use AppOnly by default,
```
,AddMicrosoftGraphAppOnly(authProvider => new GraphServiceClient(authProvider));
```

I'm still not completely satisfied with this approach because you cannot call both of these and inject them both into a controller.  We would need some way of registering the two instances separately.  But this at least gives us a place to start.

This PR also introduced the use of TokenAcquisitionProviderOption which is the class that can contain the configuration options passed on a per-request basis.,

I also renamed TokenAcquistionCredentialProvider because that mixes the names of AuthenticationProviders and TokenCredentials. As we will be moving to TokenCredentials in the future, it gets very confusing.

#654

### CODE REVIEW

**Review by baywet — APPROVED** [review]
great work! so if I understood things well, in this design people would only have one service (either app only or delegated) and locally override the context as needed. Not two services?

File: src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        /// and can be used to perform incremental scope consent.
+        /// This only works with the default authentication handler and default set of Microsoft graph authentication providers.
+        /// If you use a custom authentication handler or authentication provider, you have to handle it's retrieval in your implementation.
+        /// </summary>
+        /// <param name="baseRequest">The <see cref="IBaseRequest"/>.</param>
+        /// <param name="scopes">Microsoft graph scopes used to authenticate this request.</param>
+        public static T WithScopes<T>(this T baseRequest, string[] scopes) where T : IBaseRequest
+        {
+            string authHandlerOptionKey = typeof(TokenAcquisitionAuthenticationProviderOption).ToString();
+            AuthenticationHandlerOption authHandlerOptions = baseRequest.MiddlewareOptions[authHandlerOptionKey] as AuthenticationHandlerOption ?? new AuthenticationHandlerOption();
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs:21]: Using `[ ]` when there is nothing in the dictionary will lead to an exception: `KeyNotFoundException: The given key 'Microsoft.Identity.Web.TokenAcquisitionAuthenticationProviderOption' was not present in the dictionary`.

We'd need to test baseRequest.MiddlewareOptions.ContainsKey...

File: src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            authHandlerOptions.AuthenticationProviderOption = msalAuthProviderOption;
+            baseRequest.MiddlewareOptions[authHandlerOptionKey] = authHandlerOptions;
+
+            return baseRequest;
+        }
+
+        public static T WithAppOnly<T>(this T baseRequest, bool appOnly = true) where T : IBaseRequest
+        {
+            string authHandlerOptionKey = typeof(TokenAcquisitionAuthenticationProviderOption).ToString();
+            AuthenticationHandlerOption authHandlerOptions = baseRequest.MiddlewareOptions[authHandlerOptionKey] as AuthenticationHandlerOption ?? new AuthenticationHandlerOption();
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs:35]: Same comment as above

**Review by jmprieur — COMMENTED** [review]
Thanks so much, @darrelmiller 
here are some initial comments.

**Comment by jmprieur** [comment]
@jennyf19 
I propose that we merge this PR in a branch (different from master), and fix there the couple of comments I had.

I tested the WebAppCallsGraph and added the possibility of getting the mails (`Me.Messages`, with the "Mail.Read" scope

```CSharp
try
{
 var messages = await _graphServiceClient.Me.Messages.Request().WithScopes(new[] { "Mail.Read" }).GetAsync();
}
catch(Exception ex)
{
 // No access to mailbox
}
```

File: src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        /// <param name="baseRequest">The <see cref="IBaseRequest"/>.</param>
+        /// <param name="scopes">Microsoft graph scopes used to authenticate this request.</param>
+        public static T WithScopes<T>(this T baseRequest, string[] scopes) where T : IBaseRequest
+        {
+            string authHandlerOptionKey = typeof(TokenAcquisitionAuthenticationProviderOption).ToString();
+            AuthenticationHandlerOption authHandlerOptions = baseRequest.MiddlewareOptions[authHandlerOptionKey] as AuthenticationHandlerOption ?? new AuthenticationHandlerOption();
+            TokenAcquisitionAuthenticationProviderOption msalAuthProviderOption = authHandlerOptions?.AuthenticationProviderOption as TokenAcquisitionAuthenticationProviderOption ?? new TokenAcquisitionAuthenticationProviderOption();
+
+            msalAuthProviderOption.Scopes = scopes;
+
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs:25]: null check here on msalAuthProviderOptions and on line 26? (authHandlerOptoins)

**Review by jennyf19 — CHANGES_REQUESTED** [review]
:clock1:

**Review by jennyf19 — APPROVED** [review]
:shipit:

**Comment by jennyf19** [comment]
this is cool. i think if you know the changes to make, let's just merge to master and then make a new PR w/those changes.

---
In reply to: [713628028](https://github.com/AzureAD/microsoft-identity-web/pull/705#issuecomment-713628028) [](ancestors = 713628028)

**Pull request merged**